src/app/server-cluster/testing/TestProviderChangeListener.h

```diff
@@ -0,0 +1,41 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/data-model-provider/ProviderChangeListener.h>
+
+#include <vector>
+
+namespace chip {
+namespace Test {
+
+/// A provider change listener that stores all dirty reports
+/// into an internal vector
+class TestProviderChangeListener : public app::DataModel::ProviderChangeListener
+{
+public:
+    void MarkDirty(const app::AttributePathParams & path) override { mDirtyList.push_back(path); }
```

Comment:
Why AtrtibutePathParams rather than ConcreteAttributePath?

Reply:
That is the interface of ProviderChangeListener.

My understanding is that we can mass-mark an entire path for being dirty, specifically endpoint changing marks an entire endpoint dirty.

https://github.com/project-chip/connectedhomeip/blob/master/src/data-model-providers/codegen/CodegenDataModelProvider_Write.cpp#L226 allows endpoint vs attribute changes.

Overall the decision of the API was made by the reporting engine which took the most generic input:

https://github.com/project-chip/connectedhomeip/blob/master/src/app/reporting/Engine.cpp#L1051

Since we do have the use case of "dirty an entire larger thing" I kept that API to not make it hard to migrate.